lib/std/Build/Step/Compile.zig

```diff
@@ -246,6 +252,90 @@ pub const Kind = enum {
     @"test",
 };
 
+pub const HeaderInstallation = union(enum) {
+    file: File,
+    directory: Directory,
+
+    pub const File = struct {
+        source: LazyPath,
+        dest_rel_path: []const u8,
+
+        pub fn dupe(self: File, b: *std.Build) File {
+            // 'path' lazy paths are relative to the build root of some step, inferred from the step
+            // in which they are used. This means that we can't dupe such paths, because they may
+            // come from dependencies with their own build roots and duping the paths as is might
+            // cause the build script to search for the file relative to the wrong root.
+            // As a temporary workaround, we convert build root-relative paths to absolute paths.
+            // If/when the build-root relative paths are updated to encode which build root they are
+            // relative to, this workaround should be removed.
+            const duped_source: LazyPath = switch (self.source) {
+                .path => |root_rel| .{ .cwd_relative = b.pathFromRoot(root_rel) },
+                else => self.source.dupe(b),
+            };
```

Comment:
I tested this with a simple local project where

- package `a` exported a static library bundling build root-relative header `a.h`,
- package `b` depended on `a`, and
- package `b` built and installed an executable with a `main.c` root source file that uses symbols from `a.h` and which did both `exe.linkLibrary(lib)` and `exe.installLibraryHeaders(lib)`.

That repro failed with an "unable to update file" error prior to this workaround, but passes now.

If #19313 is accepted and implemented, this workaround can be reverted.
(Thanks @MasterQ32 for indirectly bringing this flaw to my attention.)